### DISCUSSION TITLE
`uvicorn.run` `env_file` Not Passed Correctly

### DISCUSSION THREAD

**Discussion by oliv10** [Kludex/uvicorn#2341] [Q&A]
### Initial Checks

- [ ] I confirm this was discussed, and the maintainers suggest I open an issue.
- [X] I'm aware that if I created this issue without a discussion, it may be closed without a response.

### Discussion Link

```Text
https://github.com/encode/uvicorn/discussions/2328
```

### Description

When passing a `.env` file with the `uvicorn.run` command it does not properly pass its information on to the uvicorn and FastAPI application.

### Example Code

```Python
### Test Code


import uvicorn
import fastapi
from pydantic_settings import BaseSettings
import argparse
import logging
from logging import getLogger


class Settings(BaseSettings):
    DEBUG: bool = False



SETTINGS = Settings()
app = fastapi.FastAPI()



Logger = getLogger("uvicorn")
if SETTINGS.DEBUG:
    Logger.setLevel(logging.DEBUG)
    Logger.debug("Debug Mode Enabled")
else:
    Logger.setLevel(logging.INFO)



@app.get("/")
async def root():
    if SETTINGS.DEBUG:
        return {"message": "DEBUG World"}
    else:
        return {"message": "Hello World"}


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--env-file", type=str, default=".env")
    args = parser.parse_args()
    
    uvicorn.run(app, env_file=args.env_file)
```

.env file
```bash
DEBUG=True
```

#### Incorrect Output
```python
python3 ./test_uvicorn.py

INFO:     Loading environment from '.env'
INFO:     Started server process [279760]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:40028 - "GET / HTTP/1.1" 200 OK
```

#### Correct Output
```python
uvicorn test_uvicorn:app --env-file .env

INFO:     Loading environment from '.env'
DEBUG:    Debug Mode Enabled
INFO:     Started server process [280032]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
DEBUG:    DEBUG
INFO:     127.0.0.1:60288 - "GET / HTTP/1.1" 200 OK
```

### Python, Uvicorn & OS Version

```Text
Running uvicorn 0.29.0 with CPython 3.12.2 on Linux
```

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/uvicorn/issues/2340">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/uvicorn/issues/2340/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/uvicorn/issues/2340/pledge.svg">
</picture>
</a>

**Comment by oliv10** [comment]
Maybe relates to https://github.com/encode/uvicorn/issues/1319 and https://github.com/encode/uvicorn/pull/1372.

**Comment by saparmuratx** [comment]
I faced similar problem trying to debug my FastAPI project.
uvicorn keeps ignoring specified --env-file

When uvicorn server is launched like this:
```
uvicorn src.main:app --reload --env-file src/.env 
```
with specified `src/.env`,
it seems to ignore specified env file at all and use global environment variables.

After some time trying to fix it, I have found the culprit.
VS Code automatically loads .env file from default workspace folder for Python.

![Screenshot from 2024-07-16 16-36-39](https://github.com/user-attachments/assets/cb679e8b-8e20-49f9-8878-3ce017d4b77e)

I think if terminal session has environment variables uvicorn will use them regardless of specified .env file. So after changing the .env file, restarting of terminal is required to load new values.

**Comment by hanchchch** [comment]
I know it's a bit late to answer but leaving it for whom may ran into the same problem.


The problem is that the `SETTINGS` is a global variable which is defined whenever this code is imported/executed, since it's not wrapped as a function.
```python
class Settings(BaseSettings):
    DEBUG: bool = False


SETTINGS = Settings()
app = fastapi.FastAPI()
```

Uvicorn loads the env file [here](https://github.com/encode/uvicorn/blob/master/uvicorn/config.py#L323-L327), inside the `uvicorn.run`. `SETTINGS` is being define before you call `uvicorn.run` when the env file is not loaded yet. That's why your env file is not reflected to the `SETTINGS`.

To solve this problem you need to wrap the settings variable into a function.

```python
def get_settings():
    return Settings()
```

Then, call it when it's obviously after `uvicorn.run` is already called.

```python
@app.get("/")
async def root():
    # this is invoked when the http requests arrive here.
    # thus, this can not be invoked before the server started to run.
    if get_settings().DEBUG:
        return {"message": "DEBUG World"}
    else:
        return {"message": "Hello World"}
```

Even better with dependency injection.

```python
from fastapi import Depends
from typing_extensions import Annotated

AppSettings = Annotated[config.Settings, Depends(get_settings)]

@app.get("/")
async def root(settings: AppSettings):
    if settings.DEBUG:
        return {"message": "DEBUG World"}
    else:
        return {"message": "Hello World"}
```

**Reply by 0xc18** [reply]
Thank you,
